src/UseCases/ChangeAddress/ChangeAddressUseCase.php

```diff
@@ -21,11 +21,17 @@ public function changeAddress( ChangeAddressRequest $request ): ChangeAddressRes
 			return ChangeAddressResponse::newErrorResponse( ['Unknown address.'] );
 		}
 		try {
-			$addressChange->performAddressChange( $this->buildAddress( $request ) );
+			if ( $request->isOnlyOptInDonationReceiptRequest() ) {
+				$addressChange->optOutOfDonationReceipt( $request->isOptedIntoDonationReceipt() );
+			}
+			else {
+				$addressChange->performAddressChange( $this->buildAddress( $request ) );
+			}
```

Comment:
You've missed a case  - address update *with* opt out. You can pull `optOutOfDonationReceipt` out of the `try` block and always call it. Then you also don't need an `if`.

Reply:
How come I don't need and `if`? What about the case where I don't want to perform addressChange, just the opt in?

Reply:
Ah, yeah, I was totally confused, had different aprroaches/solutions, edited the comments three times and then still wrote the wrong one.

The thing I was concerned about is distinguishing between the user managing to leave one field empty (=> error response) and leaving all address fields empty (=> success response with "opt-out only" record).

Checking just "address" in `AddressChangeRequest` is a pragmatic approach that will probably work fine in practice, especially since the client-side code will also do error checking. But im concerned that our future selves or an onboarding developer might wonder what's so special about the magic "address" field that we use it for checking if a request is opt-out-only. 

That leaves us with several approaches
1) leave a comment in `isOnlyOptInDonationReceiptRequest` to explain that we know the check is not exhaustive but should be enough for properly filled fields
2) check all fields in `isOnlyOptInDonationReceiptRequest` and write a test (with different test data) for `AddressChangeRequest` 
3) Move the empty the validation to a class, use it both in Address and the use case.
4) Add new constant `AddressChange::ADDRESS_TYPE_OPT_OUT` and use that in the request. 
5) Adding an explicit field `isOptOutOnly` to the request

At the moment I#d be in favor of 5). I'd handle the setting of that field on the client side (using a hidden field).

Reply:
>But im concerned that our future selves or an onboarding developer might wonder what's so special about the magic "address" field that we use it for checking if a request is opt-out-only.

I mean, this can easily be resolved with a short comment like "Naively checking the address field to establish if address change is only a donation receipt request". 

>At the moment I#d be in favor of 5). I'd handle the setting of that field on the client side (using a hidden field).

I am not against suggestion 5 but I feel like this is in conflict with your previous argument as it makes the code more complex and just takes additional time for very little benefit, especially considering the form is not used by that many people in relation to other pages.

I'd rather finish the ticket up today than have that change in and pull it into the next sprint is what I am saying.

Reply:
> 5.  Adding an explicit field isOptOutOnly to the request
> 
> At the moment I#d be in favor of 5). I'd handle the setting of that field on the client side (using a hidden field).

This sounds like the same solution, just extracted to the front end. I mean the check would be the same, right? If all the form fields are empty then it's just an opt in update.
I think the descriptive function name and a comment are sufficient for this.

Reply:
I just like being very explicit and clean with everything in the public API of a class instead encoding domain knowledge in comments. How the front end solves this is not a concern for this PR, where everything is about sound business logic and domain, that's why I favor 5).
But in the interest of finishing this, I'd be ok with adding a comment to the request (although for a quick finish of this PR a simple property for the request instead of a check would be even "faster" IMHO).

---

Let's have a last look at the logic here:

```
$addressChange->optOutOfDonationReceipt( !$request->isOptedIntoDonationReceipt() );
if ( ! $request->isOnlyOptInDonationReceiptRequest() ) {
				$addressChange->performAddressChange( $this->buildAddress( $request ) );
			}

```

Would cover all the cases better. As far as I can see, the current logic does not cover the case "user opts out but still gives address". But that's up the use unit tests to decide.